### ISSUE TITLE
Segfault with structure

### ISSUE THREAD

**Issue by shyuep** [atztogo/spglib#10]
Getting spacegroup with the following CIF results in a segfault at tolerances <= 1e-3. 1e-2 seems fine.

```
data_106460-ICSD
#c2014 by Fachinformationszentrum Karlsruhe, and the U.S. Secretary of
#Commerce on behalf of the United States.  All rights reserved.
_database_code_ICSD                106460
_audit_creation_date               2005/04/01
_chemical_name_systematic
;
Cobalt Erbium Gallium (5.56/8/10.44)
;
_chemical_formula_structural       'Co5.56 Er8 Ga10.44'
_chemical_formula_sum              'Co5.56 Er8 Ga10.44'
_publ_section_title
;
Er2 Rh Si3 and R2 Co Ga3 (R= Y, Tb, Dy, Ho, Er, Tm, Yb) with Lu2 Co
Ga3 type structure: new members of the Al B2 structure family
;
loop_
_citation_id
_citation_journal_abbrev
_citation_year
_citation_journal_volume
_citation_page_first
_citation_page_last
_citation_journal_id_ASTM
primary 'Journal of Alloys and Compounds' 1992 189 221 228 JALCEU
_publ_author_name
;
Gladyshevskii, R.E.;Cenzual, K.;Parthe, E.
;
_cell_length_a                     8.607(3)
_cell_length_b                     8.607
_cell_length_c                     6.898(3)
_cell_angle_alpha                  90.
_cell_angle_beta                   90.
_cell_angle_gamma                  120.
_cell_volume                       442.54
_cell_formula_units_Z              1
_symmetry_space_group_name_H-M     'P -6 2 c'
_symmetry_Int_Tables_number        190
_refine_ls_R_factor_all            0.092
loop_
_symmetry_equiv_pos_site_id
_symmetry_equiv_pos_as_xyz
  1    '-x, -x+y, -z'
  2    'x-y, -y, -z'
  3    'y, x, -z'
  4    '-x, -x+y, z+1/2'
  5    'x-y, -y, z+1/2'
  6    'y, x, z+1/2'
  7    '-x+y, -x, -z+1/2'
  8    '-y, x-y, -z+1/2'
  9    'x, y, -z+1/2'
 10    '-x+y, -x, z'
 11    '-y, x-y, z'
 12    'x, y, z'
loop_
_atom_type_symbol
_atom_type_oxidation_number
Co0+    0
Er0+    0
Ga0+    0
loop_
_atom_site_label
_atom_site_type_symbol
_atom_site_symmetry_multiplicity
_atom_site_Wyckoff_symbol
_atom_site_fract_x
_atom_site_fract_y
_atom_site_fract_z
_atom_site_occupancy
_atom_site_attached_hydrogens
_atom_site_U_iso_or_equiv
Co1 Co0+ 12 i 0.158(3) 0.316 0.043(6) 0.13 0 0.014(9)
Ga1 Ga0+ 12 i 0.158(3) 0.316 0.043(6) 0.87(7) 0 0.014(9)
Co2 Co0+ 4 f 0.3333 0.6667 0.065(6) 1. 0 0.0134(8)
Er1 Er0+ 2 b 0 0 0.25 1. 0 0.022(6)
Er2 Er0+ 6 h 0.523(1) 0.046 0.25 1. 0 0.020(5)
```

**Comment by atztogo** [comment]
Could you provide the data that can be directly put in spglib since the procedure between CIF and direct data is unclear? How much angle tolerance did you use?

**Comment by shyuep** [comment]
This is the POSCAR format:

```
Er8 Ga12 Co4
1.0
8.607000 0.000000 0.000000
-4.303500 7.453881 0.000000
0.000000 0.000000 6.898000
Er Ga Co
8 12 4
direct
0.000000 0.000000 0.750000 Er0+
0.000000 0.000000 0.250000 Er0+
0.477000 0.523000 0.750000 Er0+
0.477000 0.954000 0.750000 Er0+
0.046000 0.523000 0.750000 Er0+
0.523000 0.477000 0.250000 Er0+
0.954000 0.477000 0.250000 Er0+
0.523000 0.046000 0.250000 Er0+
0.842000 0.158000 0.957000 Ga0+
0.842000 0.684000 0.957000 Ga0+
0.316000 0.158000 0.957000 Ga0+
0.842000 0.158000 0.543000 Ga0+
0.842000 0.684000 0.543000 Ga0+
0.316000 0.158000 0.543000 Ga0+
0.158000 0.842000 0.457000 Ga0+
0.684000 0.842000 0.457000 Ga0+
0.158000 0.316000 0.457000 Ga0+
0.158000 0.842000 0.043000 Ga0+
0.684000 0.842000 0.043000 Ga0+
0.158000 0.316000 0.043000 Ga0+
0.666700 0.333400 0.935000 Co0+
0.666700 0.333400 0.565000 Co0+
0.333400 0.666700 0.435000 Co0+
0.333400 0.666700 0.065000 Co0+```

Running it with a site_tol of 1e-3 and angle_tol of 5 gives a seg fault.

**Comment by atztogo** [comment]
Thanks. But I could not reproduce your segfault. The test code is shown below. I have tested this just rinnging and also with valgrind. The versions tested are v1.9.2, current master branch, and release-1.9.4 branch. I have recently worked hard for spglib and it is in the release-1.9.4 branch. Can you see if it works or not?

```
#include "spglib.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

static int test_spg_get_dataset(void);
static int show_spg_dataset(double lattice[3][3],
                            const double origin_shift[3],
                            double position[][3],
                            const int num_atom,
                            const int types[]);
int main(void)
{
  if (!test_spg_get_dataset()) {
    printf("Symmetry search failed.\n");
    return 0;
  }
  return 1;
}

static int show_spg_dataset(double lattice[3][3],
                            const double origin_shift[3],
                            double position[][3],
                            const int num_atom,
                            const int types[])
{
  SpglibDataset *dataset;
  char ptsymbol[6];
  int pt_trans_mat[3][3];

  int i, j, size;
  const char *wl = "abcdefghijklmnopqrstuvwxyz";

  for ( i = 0; i < num_atom; i++ ) {
    for ( j = 0; j < 3; j++ ) {
      position[i][j] += origin_shift[j];
    }
  }

  dataset = spgat_get_dataset(lattice,
                              position,
                              types,
                              num_atom,
                              0.001000,
                              5.000000);
  if (dataset == NULL) {
    return 0;
  }

  printf("International: %s (%d)\n",
         dataset->international_symbol, dataset->spacegroup_number );
  printf("Hall symbol:   %s\n", dataset->hall_symbol );
  spg_get_pointgroup(ptsymbol,
		     pt_trans_mat,
		     dataset->rotations,
		     dataset->n_operations);
  printf("Point group:   %s\n", ptsymbol);
  printf("Transformation matrix:\n");
  for ( i = 0; i < 3; i++ ) {
    printf("%f %f %f\n",
	   dataset->transformation_matrix[i][0],
	   dataset->transformation_matrix[i][1],
	   dataset->transformation_matrix[i][2]);
  }
  printf("Wyckoff letters:\n");
  for ( i = 0; i < dataset->n_atoms; i++ ) {
    printf("%c ", wl[dataset->wyckoffs[i]]);
  }
  printf("\n");
  printf("Equivalent atoms:\n");
  for (i = 0; i < dataset->n_atoms; i++) {
    printf("%d ", dataset->equivalent_atoms[i]);
  }
  printf("\n");

  for (i = 0; i < dataset->n_operations; i++) {
    printf("--- %d ---\n", i + 1);
    for (j = 0; j < 3; j++) {
      printf("%2d %2d %2d\n",
	     dataset->rotations[i][j][0],
	     dataset->rotations[i][j][1],
	     dataset->rotations[i][j][2]);
    }
    printf("%f %f %f\n",
	   dataset->translations[i][0],
	   dataset->translations[i][1],
	   dataset->translations[i][2]);
  }

  spg_free_dataset(dataset);
}

static int test_spg_get_dataset(void)
{
  printf("*** Run spg_get_dataset ***:\n");
  double lattice[3][3] = {
      {   8.606999999999999,   -4.303500000000000,    0.000000000000000},
      {   0.000000000000000,    7.453881000000000,    0.000000000000000},
      {   0.000000000000000,    0.000000000000000,    6.898000000000000}
  };
  double origin_shift[3] = {0, 0, 0};
  double position[][3] =
    {
      {   0.000000000000000,    0.000000000000000,    0.750000000000000},
      {   0.000000000000000,    0.000000000000000,    0.250000000000000},
      {   0.477000000000000,    0.523000000000000,    0.750000000000000},
      {   0.477000000000000,    0.954000000000000,    0.750000000000000},
      {   0.046000000000000,    0.523000000000000,    0.750000000000000},
      {   0.523000000000000,    0.477000000000000,    0.250000000000000},
      {   0.954000000000000,    0.477000000000000,    0.250000000000000},
      {   0.523000000000000,    0.046000000000000,    0.250000000000000},
      {   0.842000000000000,    0.158000000000000,    0.957000000000000},
      {   0.842000000000000,    0.684000000000000,    0.957000000000000},
      {   0.316000000000000,    0.158000000000000,    0.957000000000000},
      {   0.842000000000000,    0.158000000000000,    0.543000000000000},
      {   0.842000000000000,    0.684000000000000,    0.543000000000000},
      {   0.316000000000000,    0.158000000000000,    0.543000000000000},
      {   0.158000000000000,    0.842000000000000,    0.457000000000000},
      {   0.684000000000000,    0.842000000000000,    0.457000000000000},
      {   0.158000000000000,    0.316000000000000,    0.457000000000000},
      {   0.158000000000000,    0.842000000000000,    0.043000000000000},
      {   0.684000000000000,    0.842000000000000,    0.043000000000000},
      {   0.158000000000000,    0.316000000000000,    0.043000000000000},
      {   0.666700000000000,    0.333400000000000,    0.935000000000000},
      {   0.666700000000000,    0.333400000000000,    0.565000000000000},
      {   0.333400000000000,    0.666700000000000,    0.435000000000000},
      {   0.333400000000000,    0.666700000000000,    0.065000000000000}
    };
  int types[] = {
      68, 68, 68, 68, 68, 68, 68, 68, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 31, 27, 27, 27, 27
  };
  int num_atom = 24;

  return show_spg_dataset(lattice, origin_shift, position, num_atom, types);
}

```

**Comment by blokhin** [comment]
With the spglib from the last commit from master, ase version 3.11, the POSCAR from above and the following script I see no problems:
```
from ase.io.vasp import read_vasp
import spglib
structure = read_vasp('/home/Evgeny/data/poscar_examples/shyue.POSCAR')
print [ (10**i, spglib.get_spacegroup(structure, symprec=10**i)) for i in range(-1, -6, -1) ]
```

**Comment by shyuep** [comment]
Where can I find spglib 1.9.4? It does not seem to be on sourceforge.

**Comment by blokhin** [comment]
@shyuep while 1.9.4 is not available directly via ```pip install spglib``` you may use the following:
```
pip install -e "git+https://github.com/atztogo/spglib.git@deb80ab901b1ad1f2df65c2eec39ca7ec4283291#egg=spglib&subdirectory=python"
```
git and pip>7.0 must be present

**Comment by atztogo** [comment]
There is  the branch release-1.9.4. This will be released soon with PyPI and conda packages.
Segfault can depend on system and very tiny difference of parameters, too. It's tough to hook it. Shyue Ping and Evgeny, I appreciate those reports and tests by others.

**Comment by atztogo** [comment]
Spglib 1.9.4 has been released. You can get it from PyPI for the python package or the source code from sourceforge. For the python interface, you don't need to use ASE Atoms class object. Instead a list of crystal structure parameters can be used. See http://atztogo.github.io/spglib/python-spglib.html#crystal-structure-cell .

**Comment by shyuep** [comment]
I confirm that this has been solved with v1.9.4.2 of spglib. Thanks.
